packages/authorizer/certora/specs/Authorizer.spec

```
@@ -1,3 +1,360 @@
+}
+
+rule revokingPermissionsIsLikeUnauthorizing(env e, address who, address where, bytes4 what)
+    good_description "Calling changePermissions with 1 revoke should be equivalent to calling unauthorize"
+{
+    storage initStorage = lastStorage;
+
+    IAuthorizer.PermissionChange change = buildPermissionChange(where, false, 0, to_bytes4(0), true, who, what);
+    changePermissions(e, [change]);
+    storage afterChangePermissions = lastStorage;
+
+    unauthorize(e, who, where, what) at initStorage;
+    storage afterUnauthorize = lastStorage;
+
+    assert afterChangePermissions[currentContract] == afterUnauthorize[currentContract];
+}
+
+rule changePermissionsInBetweenShouldNotChangeStorage
+    (env e, address who, address where, bytes4 what, IAuthorizer.Param[] params, address who2, bytes4 what2)
+    good_description "Doing authorize, change permissions, and then unauthorize shouldn't change the storage"
+{
+    require params.length == 0 && who != who2 && what != what2;
+
+    require !hasPermission(who, where, what);
+    require getPermissionParamsLength(who, where, what) == 0; // TODO: delete this line once the invarian works
+
+    require !hasPermission(who2, where, what2);
+    require getPermissionParamsLength(who2, where, what2) == 0; // TODO: delete this line once the invarian works
+
+    storage initStorage = lastStorage;
+
+    authorize(e, who, where, what, params);
+
+    IAuthorizer.PermissionChange change = buildPermissionChange(where, true, who2, what2, true, who, what);
+    changePermissions(e, [change]);
+
+    unauthorize(e, who2, where, what2);
+
+    storage currentStorage = lastStorage;
+    assert initStorage[currentContract] == currentStorage[currentContract];
+}
+
+rule changePermissionsOrderMatters(env e, address who, address where, bytes4 what)
+    good_description "Calling changePermissions with a grant and its corresponding revoke should leave the user unauthorized"
+{
+    require !hasPermission(who, where, what);
+    require getPermissionParamsLength(who, where, what) == 0; // TODO: delete this line once the invarian works
+
+    storage initStorage = lastStorage;
+
+    IAuthorizer.PermissionChange change = buildPermissionChange(where, true, who, what, true, who, what);
+    changePermissions(e, [change]);
+
+    storage currentStorage = lastStorage;
+
+    assert !hasPermission(who, where, what);
+    assert initStorage[currentContract] == currentStorage[currentContract];
+}
+
+rule hasPermissionIsEquivalentToIsAuthorizedWithEmptyParams(address who, address where, bytes4 what)
+    good_description "For a permission without params, hasPermission should be equivalent to isAuthorized"
+{
+    require getPermissionParamsLength(who, where, what) == 0;
+    requireNonGenericPermissions(who, where, what);
+
+    uint256[] how = helpers.authParams(who, where, what);
+    assert isAuthorized(who, where, what, how) == hasPermission(who, where, what);
+}
```

It should work for any how:

```suggestion
rule hasPermissionIsEquivalentToIsAuthorizedWithEmptyParams(address who, address where, bytes4 what, uint256[] how)
    good_description "For a permission without params, hasPermission should be equivalent to isAuthorized"
{
    require getPermissionParamsLength(who, where, what) == 0;
    requireNonGenericPermissions(who, where, what);

    assert isAuthorized(who, where, what, how) == hasPermission(who, where, what);
}
```